rpc/src/connection.rs

```diff
@@ -0,0 +1,540 @@
+        let mut header = [0; RESPONSE_HEADER_SIZE];
+        self.receive(&mut header)?;
+        debug!("header: {:?}", header);
+        let id = u32::from_le_bytes([header[0], header[1], header[2], header[3]]);
+        let status = i32::from_le_bytes([header[4], header[5], header[6], header[7]]);
+        let flags = u32::from_le_bytes([header[8], header[9], header[10], header[11]]);
+        let total_length = u32::from_le_bytes([header[12], header[13], header[14], header[15]]);
+        let meta_data_length = u32::from_le_bytes([header[16], header[17], header[18], header[19]]);
+        let data_length = u32::from_le_bytes([header[20], header[21], header[22], header[23]]);
+        debug!(
+            "id: {}, status: {}, flags: {}, total_length: {}, meta_data_length: {}, data_length: {}",
+            id, status, flags, total_length, meta_data_length, data_length
+        );
+        Ok(ResponseHeader {
+            id,
+            status,
+            flags,
+            total_length,
+            meta_data_length,
+            data_length,
+        })
+    }
+
+    pub fn receive_response(
+        &self,
+        data: &mut [u8],
+        meta_data: &mut [u8],
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let data_length = data.len();
+        let meta_data_length = meta_data.len();
+        // if data_length > MAX_DATA_LENGTH.try_into().unwrap()
+        //     || meta_data_length > MAX_METADATA_LENGTH.try_into().unwrap()
+        // {
+        //     return Err("data length or meta data length is too long".into());
+        // }
+        self.receive(&mut data[0..data_length as usize])?;
+        self.receive(&mut meta_data[0..meta_data_length as usize])?;
+        Ok(())
+    }
+
+    pub fn receive(&self, data: &mut [u8]) -> Result<(), Box<dyn std::error::Error>> {
+        let mut stream = self.stream.as_ref().unwrap();
+        let mut buf_len = 0;
+        debug!("waiting for response, data length: {}", data.len());
+
+        // TODO: use epoll or async read
+        while buf_len < data.len() {
+            let result = stream.read(data);
+            match result {
+                Ok(len) => {
+                    buf_len += len;
+                    debug!("received {} bytes, total: {}", len, buf_len);
+                }
+                Err(_) => {
+                    return Err("failed to receive response".into());
+                }
+            }
+            sleep(Duration::from_millis(100));
+        }
+        debug!("received response, data length: {}", buf_len);
+        Ok(())
+    }
+
+    pub fn clean_response(&self, total_length: u32) -> Result<(), Box<dyn std::error::Error>> {
+        let mut buffer = vec![0u8; total_length as usize];
+        self.receive(&mut buffer)?;
+        Ok(())
+    }
+}
+
+pub struct ServerConnection {
+    //stream: Arc<tokio::sync::Mutex<TcpStream>>,
+    write_stream: Arc<tokio::sync::Mutex<OwnedWriteHalf>>,
```

Comment:
Why not keep using ‘_send_lock’, and I don't think it's necessary to use `Arc`

Reply:
I have no idea about how to send a mutable reference between coroutine without using tokio::sync::mutex, when I changed std::net::tcpstream to tokio::net::tcpstream, even use raw pointer (It need the stream struct be created on the heap).

Reply:
> Why not keep using `_send_lock`, and I don't think it's necessary to use `Arc` because of `Arc<ServerConnection>`

Maybe I misunderstood your comment. I'll fix it.

Reply:
Done.